Fix 483371 : client hello after client handshake crash

sophokles73 opened the pull request:
i know, it's a lot of changes but after having split up the record processing code into multiple much smaller methods I also wanted to get rid of declaring checked HandshakeException on every single one of the methods ...

sbernard31 committed:
Fix Bug 483371.

- Add unit test for a client which crashes during handshake
- Add test case to check re-transmission of CLIENT_HELLO doesn't restart
ongoing handshake
- Incoming handshake messages are now matched against ongoing handshake
and/or established session.
- CLIENT_HELLO from epoch 0 always leads to client IP verification
(HELLO_VERIFY) and new handshake being started (discarding any ongoing
handshake) if successful.
- Re-transmitted (duplicate) CLIENT_HELLO messages are matched against
ongoing handshake in order to prevent restarting the handshake

Also-by: Kai Hudalla <kai.hudalla@bosch-si.com>

jvermillard commented on scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java:
```diff
@@ -401,44 +406,56 @@ private void receiveNextDatagramFromNetwork() throws IOException {
+						StringBuilder msg = new StringBuilder("Aborting handshake with peer [")
+								.append(record.getPeerAddress()).append("]: ");
+						if (LOGGER.isLoggable(Level.FINEST)) {
+							msg.append(System.lineSeparator()).append(e);
```
this will not display the stacktrace.
anyway I wonder what is the purpose of so much logger code :)

sophokles73 replied:
The idea is to only be more verbose if logging is set to a finer level ...

jvermillard commented on scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java:
```diff
@@ -506,41 +537,47 @@ private void terminateConnection(Connection connection, AlertMessage alert, DTLS
+						throw getDtlsHandshakeException(e, record.getPeerAddress());
+					} catch (GeneralSecurityException e) {
+						throw new DtlsHandshakeException(
+								String.format("Cannot de-crypt %s message from peer", record.getType()),
```
"decrypt" and not "de-crypt"?

sophokles73 replied:
a matter of taste, I think. Will change to "decrypt" (because it's your birthday ;-))

jvermillard replied:
I don't think it's a matter of taste :trollface:  https://dictionary.cambridge.org/spellcheck/english/?q=de-crypt

jvermillard commented on scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java:
```diff
@@ -610,253 +653,356 @@ private void processAlertRecord(InetSocketAddress peerAddress, Record record) th
+		// by exchanging a cookie as described in section 4.2.1 of the DTLS 1.2 spec
+		// see http://tools.ietf.org/html/rfc6347#section-4.2.1
+		byte[] expectedCookie = generateCookie(clientHello);
+		if (!Arrays.equals(expectedCookie, clientHello.getCookie())) {
```
return true/false?
This is a potential DoS issue: http://java-performance.info/throwing-an-exception-in-java-is-very-slow/
Exception are slow in general, not a good idea to use it for something else than unfrequent code path

sophokles73 replied:
This is great feedback, @jvermillard. Thanks a lot for pointing this out. I will change that to returning a boolean instead ...

sbernard31 commented:
You're right there is a lot of change here. Maybe a bit risky to change so much code to fix a bug we should integrate in a 1.0.1. But I suppose "in our tests, trust we must"

My comments :
1) I like we have now smaller method.
2) I not really like all this new exception throwing. This is probably less efficient and I'm not sure this makes the code more readable.
3) This fix the problem, but now I think a CLIENT_HELLO retransmission could restart an ongoing handshake that why I propose to use the "Random" of CLIENT_HELLO to make the difference between new CLIENT_HELLO and potential retransmission. (see [code](https://github.com/eclipse/californium.scandium/blob/d4b094964c4bf64819ddea0499f2bb43927c0e4a/scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java#L745))

sophokles73 commented:
@sbernard31,

1) I also like the (now) smaller methods the most :-)
2) Based on the comment Julien made, I think that it would actually be better to catch the checked exceptions and simply log the problem if the handshake does not need to be aborted. However, I really want to get rid of the checked exceptions altogether. I would like the `Handshaker` hierarchy to throw unchecked exceptions only in the future. We could do this either by turning `HandshakeException` into an unchecked exception or by throwing `DtlsHandshakeException` instead. For the time being it seemed easier to introduce a new unchecked exception hierarchy based on `DtlsException`.
3) This is a very precise and correct observation :-) Would you mind adding a separate test case for this as well?

I think we should add commits to this branch and then later consolidate with an interactive rebase,
 ok?

sbernard31 commented:
ok I will write a test for that :), but  _"If you require from everybody that submits a bug to write an automatic test case for it, then nobody will submit bug requests."_ :p

sophokles73 commented:
I will never accept *requests* for bugs anyway ;-) Accepting bug *reports* is another thing ...

Commit:
Throw HandshakeException when parsing malformed alert message.

Throws DECODE_ERROR exception when parsing message containing illegal
values for either alert level or alert description.

Commit:
Do not throw checked exception from constructor anymore.

Throw IllegalStateException instead of HandshakeException in constructor
if JVM does not provide a SHA-256 message digest implementation.

This, however, is very unlikely because this digest is mandatory for any
Java SE 7 VM to implement.

Commit:
Turn into utility class by making default constructor private.

Commit:
Change test category to medium.

sophokles73 commented:
I have incorporated your proposed changes and would like to push them to this branch. However, I think it would be easier for you to review if I would do a rebase first and squash some commits and bring them into the right order. Any objections?

sbernard31 commented:
No objection.
I commit the test about CLIENT_HELLO retransmission. (I test it with my branch to be sure the test is ok)

sophokles73 commented:
The first four commits are for convenience only and should probably be pushed to master independently ...

sophokles73 commented:
@sbernard31, WDYT?

sbernard31 commented:
Except this 2 remarks seems ok :)

sophokles73 commented:
@sbernard31, I have incorporated the proposed changes... if you have no further comments I would rebase on master and merge ...

sophokles73 commented:
I think this is it now...

sbernard31 commented:
seems good
